### DISCUSSION TITLE
MPI and OpenMP in WarpX / How to alert WarpX about available resources on HPC?

### DISCUSSION THREAD

**Discussion by budjensen** [BLAST-WarpX/warpx#4751] [Q&A]
I have WarpX installed two different ways--from source and from conda--on the Stellar HPC at Princeton/PPPL (information on the cluster [HERE](https://researchcomputing.princeton.edu/systems/stellar)) and am just getting started running capacitive discharge simulations. I have the example python input script and run it (within a slurm script) as follows:
```
srun -N 1 -n 16 python PICMI_inputs_1d.py -n 1
```

As a primer, when I run from my own build, I see the following lines at startup:
```
Initializing AMReX (24.02-30-g2ecafcff4013)...
MPI initialized with 1 MPI processes
MPI initialized with thread support level 3
OMP initialized with 4 OMP threads
```
whereas running Warpx installed from the conda distribution displays:
```
Initializing AMReX (24.01)...
OMP initialized with 4 OMP threads
```
Is this expected behavior, that the conda distribution does not use MPI? (In contrast, my build from source had the `WarpX_MPI` option ON.)

--------------

Secondly, I submit a batch script that looks like:

```bash
#!/bin/bash
#SBATCH -J Warp_Turn
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=16
#SBATCH --cpus-per-task=4
#SBATCH --time=00:30:00
#SBATCH --output warpx.%j.out

export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK

# load conda environment
module purge
module load anaconda3/2023.9
conda activate warpx

srun -N 1 -n 16 python PICMI_inputs_1d.py -n 1
```

I have worked out that setting `OMP_NUM_THREADS` determines the number of threads initialized in the startup line `OMP initialized with 4 OMP threads`, but have not had the number in the line `MPI initialized with 1 MPI processes` move beyond 1, even when submitting a job to more than one core and using the flag `-n <num_cores>`. Instead, when submitting a job with `--ntasks-per-node=N` I find that the MPI/OMP startup lines are printed out _N_ times. This makes me think that MPI is not utilizing the full system, and perhaps even is running _N_ identical simulations in parallel. How can I alert WarpX that I have 16 MPI tasks allocated for my job?

As an additional clarification on a high level, is it correct that WarpX (via AMReX) uses MPI to distribute the simulation domain across processors and then uses OpenMP to accelerate each process?

Thank you!

**Comment by ax3l** [comment]
Usually, you would remove `-N 1` (and even `-n`) from `srun` and handle the number of processes from the `#SBATCH` details passed to the job script.

Let's continue this in https://github.com/ECP-WarpX/WarpX/issues/4752

**Comment by budjensen** [comment]
Sure thing, thanks
